### PR TITLE
Add `second_factors`

### PR DESCRIPTION
**Pull request by Joerger**
Add `second_factors` and deprecate `second_factor`

We don't currently plan on removing `second_factor`, as this would require a more complicated migration process. Instead we will just derive `second_factors` from `second_factor` and output a warning log when `second_factor` is set. 

In this PR I've also added the SSO second_factor type. It is currently completely unused, but we'd rather get the proto changes into v17 rather than waiting until SSO MFA is fully released in a minor version.

Note: I have not propogated the `second_factors` change to the Teleport Connect or WebUI, meaning those still rely on the `second_factor` from proxy ping response. I will do this in a follow up PR. 

Based off https://github.com/gravitational/teleport/pull/47153 due to reduced dependence on second_factor from proxy ping response.

TODO: I expect some test breakages, will fix after opening. Also cleanup.

Follow up TODO: Update docs.

Changelog: Add new `second_factors` field to cluster auth preference for more clarity and granularity over which 2fa methods are enabled in a cluster.

### CODE REVIEW

File: api/constants/constants.go [hunk 1]
```diff
@@ -191,6 +191,7 @@ var SystemConnectors = []string{
 }
 
 // SecondFactorType is the type of 2FA authentication.
+// Deprecated: Use types.SecondFactorType
```
Comment (codingllama) [inline api/constants/constants.go:194]: Will this cause problems on e/ ?

(I'm not sure deprecating this is wise, see other comment threads.)

File: api/proto/teleport/legacy/types/types.proto [hunk 2]
```diff
@@ -2099,6 +2099,20 @@ message AuthPreferenceSpecV2 {
   // SignatureAlgorithmSuite is the configured signature algorithm suite for the cluster.
   // The current default value is "legacy". This field is not yet fully supported.
   SignatureAlgorithmSuite signature_algorithm_suite = 20;
+
+  // SecondFactors is a list of supported second factor types.
```
Comment (codingllama) [inline api/proto/teleport/legacy/types/types.proto:2103]:
```suggestion
  // SecondFactors is a list of supported second factor types, in ascending
  // order of preference (first item is preferred).
```

File: api/constants/constants.go
Comment (codingllama) [inline api/constants/constants.go:1]: > and deprecate second_factor

Do we have to? Every config out there uses it, plus I see little harm in keeping it around.

File: integration/db/fixture.go [hunk 3]
```diff
@@ -293,7 +294,7 @@ func SetupDatabaseTest(t *testing.T, options ...TestOptionFunc) *DatabasePack {
 	rcConf := servicecfg.MakeDefaultConfig()
 	rcConf.DataDir = t.TempDir()
 	rcConf.Auth.Enabled = true
-	rcConf.Auth.Preference.SetSecondFactor("off")
+	rcConf.Auth.Preference.SetSecondFactor(constants.SecondFactorOff)
```
Comment (codingllama) [inline integration/db/fixture.go:297]: Taking the PR as-is, isn't this introducing a new usage of a deprecated constant?

**Review by codingllama — COMMENTED** [review]
Taking a quick look, I'm not certain we should deprecate second_factor (see comment threads).

If we do go that way, I think we should make this PR only about adding second_factors, and follow up with a deprecation and widespread changes separately. It helps greatly in bounding the review.

**Comment by Joerger** [comment]
> Taking a quick look, I'm not certain we should deprecate second_factor (see comment threads).

We don't plan on removing `second_factor` any time soon. We will prefer if `second_factors` is set, but derive it from `second_factor` otherwise for the foreseeable future.

> If we do go that way, I think we should make this PR only about adding second_factors, and follow up with a deprecation and widespread changes separately. It helps greatly in bounding the review.

I will try to split this into a couple PRs.

File: api/types/enum.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import (
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+// decodeEnum decodes a protobuf enum from a representational value, usually a bool,
+// string, or from the actual enum (int32) value. If the value is valid, it is saved
+// in the given enum pointer.
+func decodeEnum[T ~int32](p *T, val any, representationMap map[any]T, enumMap map[int32]string) error {
```
[the remark this answers is not in the document]
Comment (Joerger) [reply api/types/enum.go:28]: https://github.com/gravitational/teleport/pull/47230

**Comment by codingllama** [comment]
Thanks, Brian!

**Pull request closed without merging**